### DISCUSSION TITLE
How to use custom types inside Model?

### DISCUSSION THREAD

**Discussion by MuhannadAlrusayni** [SeaQL/sea-orm#139] [General]
Hi, I want to use my custom types in the Model:
```rust
#[derive(Clone, Debug, PartialEq, DeriveModel, DeriveActiveModel)]
pub struct Model {
    pub id: AccountId,
    pub username: Username,
    pub email: Email,
    pub password: String,
    pub salt: String,
    pub created_at: DateTime,
    pub locked: Option<String>,
    pub disabled: Option<String>,
}
```
There are a few issues I faced while I'm trying to do so:
- Custom types support was not mentioned at all in the [docs](https://www.sea-ql.org/SeaORM/docs/index), [SeaORM API docs](https://docs.rs/sea-orm/0.1.2/sea_orm/index.html), [SeaQuery API docs](https://docs.rs/sea-query/0.15.0/sea_query/index.html) and the examples. the only place it was mentioned is #107 
- I was not able to implement `ValueType` for my custom types because it requires `ValueTypeDefault` that's not implemented for all of my custom types (e.g. you cannot have default impl for `Email`)

Another questions that arise in my head is how SeaORM would be able to convert the row table to my `Model` back since there is no information that tells SeaORM how to do so, so I think we need a way to do so, for example `TryFrom<Value> for Email` or `FromSqlType` trait?

Another thing that I notice is, do we really need `ValueTypeDefault`? why not just use `Default` trait since it's already bounded:
```rust
impl ValueTypeDefault for $type
where
    $type: Default,
{
    fn default() -> Self {
        Default::default()
    }
}
```
And I'm not sure if we going to keep this bound, since it's not possible to impl `Default` for all custom types.

Sorry for the long issue, but I think this is an important feature to have in an ORM. and yeah this ORM look very promising, keep up the good work

**Comment by billy1624** [comment]
Hi @MuhannadAlrusayni, thanks for raising this issue. Yes, we do not support custom type defined as above, however we do support `Decimal`, `Uuid`, etc. third party types.

But I do think we need to open up the API to support all custom types in general. 

Can I know what is the type definition of `Email`? Is it a struct with multiple attributes inside it?

**Comment by tyt2y3** [comment]
I think you have made a good point, `ValueTypeDefault` seemingly is not needed now.
I've removed it in https://github.com/SeaQL/sea-query/pull/114

**Comment by MuhannadAlrusayni** [comment]
> Can I know what is the type definition of Email? Is it a struct with multiple attributes inside it?

It's just newtype with constraint on the value in side (aka [ValueObject](https://en.wikipedia.org/wiki/Value_object)):
```rust
#[derive(Display, SerializeDisplay, DeserializeFromStr)]
pub struct Email(String);

impl FromStr for Email {
    type Err = Error;

    fn from_str(s: &str) -> Result<Self> { .. }
}
```
and I use these types in my endpoints API to get user input, this way I can make sure values passed to my backend are valid data.

**Comment by MuhannadAlrusayni** [comment]
Thanks @tyt2y3 using https://github.com/SeaQL/sea-query/pull/114 I was able to use my custom types. It would be nice if we have this functionality in`sea-orm` as soon is you can :D. 

Edit: Another things that's out of this issue scope, would it be possible to expose `type_to_value` macro or even better derive macro for newtypes to do the mapping? because I end up having almost an exact copy of `type_to_value` macro in my code base.

**Comment by tyt2y3** [comment]
Okay great. So what are the missing pieces for now? Only a version bump of SeaQuery inside SeaORM?
________________________________
寄件者: Muhannad ***@***.***>
寄件日期: Tuesday, August 31, 2021 6:37:29 AM
收件者: SeaQL/sea-orm ***@***.***>
副本: Chris Tsang ***@***.***>; Mention ***@***.***>
主旨: Re: [SeaQL/sea-orm] How to use custom types inside Model? (#116)


Thanks @tyt2y3<https://apac01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ftyt2y3&data=04%7C01%7C%7C875322db03b44aaa78db08d96c06c06f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637659598525701909%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=uoRFlwE6EwyOAwR5eMMGKiXtRKLbPAGq6qSwCqhjazI%3D&reserved=0> using SeaQL/sea-query#114<https://apac01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FSeaQL%2Fsea-query%2Fpull%2F114&data=04%7C01%7C%7C875322db03b44aaa78db08d96c06c06f%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C637659598525711865%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=6Pu6MSQdUt5I6G8Hwhkp7PjoFwEyT%2BFzmFZD1HQZfdk%3D&reserved=0> I was able to use my custom types. It would be nice if we have this functionality insea-orm as soon is you can :D.

**Comment by MuhannadAlrusayni** [comment]
I'm not sure yet, but I think version bump all what we need.

**Comment by tyt2y3** [comment]
Great. We will do so soon.

**Comment by tyt2y3** [comment]
Will you try `0.2.0` works for you?
